Form variables

jeremicmilan opened the pull request:
Form variables are a feature to configure the variables you want to be forwarded from scraping the form-submit page to scraping the main page for sensor data. A common use case is to populate the `X-Login-Token` header which is the result of the login.

Example: 
```yaml
multiscrape:
  - resource: "https://website-api.airvisual.com/v1/users/65a28a0cec1ff309a74ba414/devices/avo_65a2bd77c2f7aeabcd715393?units.system=metric&AQI=US&language=en"
    form_submit:
      submit_once: True
      resource: "https://website-api.airvisual.com/v1/auth/signin/by/email"
      input:
        email: "<email>"
        password: "<password>"
      variables:
        - name: token
          value_template: "{{ (value | from_json).loginToken }}"
    headers:
      X-Login-Token: "{{ token }}"
    sensor:
      - name: AirVisual Outdoor AQI
        value_template: "{{ (value | from_json).current.aqi.value }}"
        unit_of_measurement: "AQI US"
```

Log into `https://dashboard.iqair.com/personal/devices`, select the device to get the `<device_id>` in the URL. After that analyze network traffic and find the name starting with `<device_id>`. That will contain the entire path in the example including `<user_id>` (there's probably an easier way to get `<user_id>`, but this works),

jeremicmilan commented:
This is continuation of #327. Please refer to it for history.

jeremicmilan commented:
@danieldotnl, debugging the extension does not work for me. It worked when I initially worked on the first PR (header mappings). Does it work for you?

My setup is WSL, I enable debugpy and attach locally. The debugger attaches, but it refuses to hit breakpoints (where before it did). Very weird. Do you know what's going on?
BTW, I implemented the form variables without the debugger, it was a little bit annoying. :D

danieldotnl commented on README.md:
```diff
@@ -135,18 +131,14 @@ Configure a refresh button to manually trigger scraping.
+| name | description                                   | required | default | type   |
+| ---- | --------------------------------------------- | -------- | ------- | ------ |
+| name | Name of the attribute (will be slugified)     | True     |         | string |
+|      | Shared fields from the [Selector](#Selector). | True     |         |        |
```
'See Selector fields' would be more in line with the other links.

danieldotnl commented on README.md:
```diff
@@ -156,6 +148,50 @@ Configure the form-submit functionality which enables you to submit a (login) fo
+
+### Form Variables
+
+Configure the variables you want to be forwarded from scraping the [Form-submit](#form-submit) page to scraping the main page for sensor data. A common use case is to populate the `X-Login-Token` header which is the result of the login.
```
I propose:
Configure the variables that will be scraped from the `form_submit` response. You  will be able to use those values in the `value_template` of a `header` or a `selector` in the main configuration. A common use case is to populate the `X-Login-Token` header which is the result of the login.

danieldotnl replied:
I see now that you also use the variables in the request data and params templates. Let's mention that as well here.

jeremicmilan replied:
I'm unsure how to phrase this. Can you suggest, please?

jeremicmilan replied:
Let's continue this discussion in the other comment.

danieldotnl commented on README.md:
```diff
@@ -156,6 +148,50 @@ Configure the form-submit functionality which enables you to submit a (login) fo
+| name | description                                   | required | default | type   |
+| ---- | --------------------------------------------- | -------- | ------- | ------ |
+| name | Name of the variable                          | True     |         | string |
+|      | Shared fields from the [Selector](#Selector). | True     |         |        |
```
See selector fields.. (like the comments above)

danieldotnl commented on README.md:
```diff
@@ -156,6 +148,50 @@ Configure the form-submit functionality which enables you to submit a (login) fo
+|      | Shared fields from the [Selector](#Selector). | True     |         |        |
+
+Example:
+
```
I don't like to use a specific (commercial) site here as an example. Could you create an anonymous example. Like: resource: "https://somesiteyouwanttoscrape.com"

danieldotnl commented on README.md:
```diff
@@ -156,6 +148,50 @@ Configure the form-submit functionality which enables you to submit a (login) fo
+
+### Selector
+
+Shared field used in multiple configs above. Used to define the scraping: how to extract a value from the page.
```
Used to configure scraping options

danieldotnl commented on README.md:
```diff
@@ -176,7 +212,7 @@ Multiscrape also offers a `get_content` and a `scrape` service. `get_content` re
 
 Both services accept the same configuration as what you would provide in your configuration yaml (what is described above), with a small but important caveat: if the service input contains templates, those are automatically parsed by home assistant when the service is being called. That is fine for templates like `resource` and `select`, but templates that need to be applied on the scraped data i ...
-*If someone has a better solution, please let me know!*
+_If someone has a better solution, please let me know!_
```
This is the same, isn't it? :)

jeremicmilan replied:
Yes, but it was done automatically. Not sure if the new gates you introduced or one of extensions I have installed.

danieldotnl commented on custom_components/multiscrape/http.py:
I rather have the variables as state in the coordinator instead of in the HttpWrapper class.  So to provide the variables as parameter to `async_request`, similar to what I did with cookies in the other PR.

jeremicmilan replied:
Makes sense. Done.

danieldotnl commented on custom_components/multiscrape/scraper.py:
Similar to what I wrote for the HttpRequest, ideally I have the variables just as state in the coordinator. However, that's complicated since the scrape function is called from by the sensor class.
So either we store the variables in hass data or we leave it like this, but then we should clean up the variables in the reset function in the scraper.

jeremicmilan replied:
No need to add additional things in hass data, as coordinator is already visible from entity/sensor classes. Note that both you and myself are storing the data in request manager which is a field in the coordinator. Hope that you like this approach more.

jeremicmilan replied:
Added a getter get_form_variables in the coordinator.

jeremicmilan replied:
Reactivating comment as I should define a property and not a getter function.

jeremicmilan replied:
Changed to a property.

danieldotnl commented on custom_components/multiscrape/util.py:
```diff
@@ -5,23 +5,22 @@
         value_template.hass = hass
 
-    def _render(value: Any = None, parse_result=False):
+    def _render(value: Any = None, parse_result=False, variables: dict = {}):
```
Would it be cleaner to rename the `value` parameter to `values` or `variables` and check if it is a dictionary? If not, keep old behavior, other just provide variables and assume the dictionary has a value.
Your solution is more explicit but value and variables are actually both variables. What do you think?

jeremicmilan replied:
I decided to go with the option to rename value to variables, as I couldn't find usages of `value`. I guess it used before and usage stopped after a while.

danieldotnl reviewed (changes requested):
Very nice that you picked this up and implemented variables! Thank you!

jeremicmilan commented on requirements.txt:
```diff
@@ -1,4 +1,4 @@
 colorlog==6.8.2
-pytest-homeassistant-custom-component==0.13.111
+pytest-homeassistant-custom-component==0.13.133
```
@danieldotnl , please sign off this change. Previous version did not build the container for me as it could not install mypy-dev==1.10.0a3

danieldotnl replied:
I already merged that in #372

jeremicmilan replied:
Awesome. Merged with latest.

jeremicmilan commented:
@danieldotnl, it took a while to get some time for this, but I finally addressed all the feedback. Please review again.

jeremicmilan commented on custom_components/multiscrape/coordinator.py:
```diff
@@ -50,12 +46,15 @@ def __init__(
         self._http = http
         self._form_submitter = form
         self._resource_renderer = resource_renderer
+        self._scraper = scraper
```
Not needed anymore

jeremicmilan commented on custom_components/multiscrape/scraper.py:
```diff
@@ -45,6 +45,7 @@ def __init__(
         self._soup: BeautifulSoup = None
         self._data = None
         self._separator = separator
+        self._variables = {}
```
Not used anymore

jeremicmilan commented on custom_components/multiscrape/util.py:
```diff
@@ -19,14 +17,14 @@ def create_renderer(hass, value_template):
+                "Error rendering template: %s with variables %s", value_template, variables
             )
-            return value
+            return None
```
@danieldotnl, I'm returning None now. Is that ok?

danieldotnl replied:
I think we should re-raise the exception as we are in trouble anyway when this happens.

jeremicmilan replied:
Makes sense.

jeremicmilan commented:
@danieldotnl, fun trivia, the token that I started using months ago did not yet expire. It usually should expire in 90 days. So, I did not need to update manually the token on my production home assistant installation. 😄

danieldotnl commented on custom_components/multiscrape/__init__.py:
```diff
@@ -117,22 +106,25 @@ async def _async_process_config(hass: HomeAssistant, config) -> bool:
-            form_http = create_http_wrapper(config_name, form_submit_config, hass, file_manager)
             parser = conf.get(CONF_PARSER)
+            form_http = create_http_wrapper(config_name, form_submit_config, hass, file_manager)
+            form_scraper = create_scraper(config_name, conf, hass, file_manager)
```
Let's only create this one when variables are given in the config. Otherwise set it to None.

jeremicmilan replied:
Moved the logic to `create_form_submitter` and done that.

danieldotnl commented on custom_components/multiscrape/form.py:
```diff
@@ -150,11 +154,20 @@ async def async_submit(self, main_resource):
         if self._submit_once:
             self._should_submit = False
 
+        await self._scraper.set_content(response.text)
```
Only required in case of variables, right?

jeremicmilan replied:
Yes. Per other comment we are not initializing scraper if there are no variables, so checking here if scraper is initialized. It's bettercheck anyways, if we decide to use scraper in the form submit for something else other than variables.

danieldotnl commented on custom_components/multiscrape/service.py:
```diff
@@ -151,7 +140,8 @@ async def _prepare_service_request(hass: HomeAssistant, conf, config_name):
+        form_http = create_http_wrapper(
+            config_name, form_submit_config, hass, None)
         form_submitter = create_form_submitter(
             config_name, form_submit_config, hass, form_http, None, parser
```
You need to create a scraper and provide it here in case variables are configured. Did you test the service?

jeremicmilan replied:
Moved scraper creation to the `create_form_submitter` function, as it's better suited there.

jeremicmilan replied:
About testing the service... I did not... Testing it now, the example you put is not working anymore (even on master branch, and even the default configuration file) so I'm unsure what's working properly and what's not working properly in the service. Also, the coordinator is not initialized for sensor's under service, so I'm unsure how to handle form variables in the service at all. The logic I've added in service.py I currently made safe by checking if coordinator is initialized (which fixed an exception caused by that), but I can remove changes entirely in the service.py, whatever you prefer. Currently I do not need the service, working with form variables, so I'm fine with them not working. Waiting on your guidance. Maybe you fix the current aster branch and then I merge that and follow-up afterwards?
I would look into what's happening in more detail, but debugging does not work for me. The debugger attaches, but it does not hit breakpoints (it did when I initialy implemented the first version - header mappings). Any idea what's up with that?

danieldotnl replied:
The example was not working because home-assistant.io was redesigned recently 😅
I updated the examples in the readme now.
I will also have a look at how to implement the variables in the service, as I find it important that there is feature parity.

jeremicmilan replied:
Thanks for fixing the README. You should update the example `configuration.yaml` as well.

Good news: I have service working for form variables! :)
The main issue is the custom handling of templates by multiscrape (`!` within `{{` and `}}`). Please look at the latest changes.

jeremicmilan replied:
Thanks for fixing the README. You should update the example `configuration.yaml` as well.

Good news: I have service working for form variables! :)
The main issue is the custom handling of templates by multiscrape (`!` within `{{` and `}}`). Please look at the latest changes.

danieldotnl reviewed (changes requested):
It's taking some time but we are getting close!!

danieldotnl commented:
> @danieldotnl, fun trivia, the token that I started using months ago did not yet expire. It usually should expire in 90 days. So, I did not need to update manually the token on my production home assistant installation. 😄

Good for you as this took way longer than 90 days 😅

danieldotnl commented:
Let's try to finish this soon now, as people are waiting for the cookies PR to be merged, but I don't want to bother you with merging that one as well in this PR. So I would like to merge you PR first.

jeremicmilan commented:
@danieldotnl , I'll have a lot more time in July. Probably won't make it before that.

jeremicmilan commented on README.md:
```diff
@@ -156,6 +148,47 @@ Configure the form-submit functionality which enables you to submit a (login) fo
+
+### Form Variables
+
+Configure the variables that will be scraped from the [`form_submit`](#form-submit) response. You will be able to use those values in the `value_template` of a header or a selector in the main configuration. A common use case is to populate the `X-Login-Token` header which is the result of the login.
```
ok, I might have blindly made those changes without testing them. :D I'll test them out and report back.

jeremicmilan replied:
Tested sensor and attribute configuration and they work. Updated README accordingly. Were you thinking about these with 'request data' and 'params' templates? Or something else entirely.

jeremicmilan replied:
Tested sensor and attribute configuration and they work. Updated README accordingly. Were you thinking about these with 'request data' and 'params' templates? Or something else entirely?

jeremicmilan commented on README.md:
```diff
@@ -156,6 +148,47 @@ Configure the form-submit functionality which enables you to submit a (login) fo
+        - name: token
+          value_template: "{{ ... }}"
+    headers:
+      X-Login-Token: "{{ token }}"
```
@danieldotnl, when using variables, should we use them directly or should we add a prefix? For example: `form_variables.token`. This is basically wrapping all form variables into one json object/dict which should be more future proof. On the other hand, it might be overengineering. :D

danieldotnl replied:
It's a very good thought, but indeed also over-engineering and adding a bit of extra complexity for the user. I think when we have more types of variables in the future, we should add a check for duplicate variable names instead. So leave it like it is for now.

jeremicmilan commented:
@danieldotnl, I addressed comments and I'm waiting for your guidance on some of them. Please review.

Also, I would love to add tests for these cases, so as they do not get broken over time. I tried adding a test and did not succeed after half an hour. I saw you doing stuff in that space recently. Do you plan to add tests (examples I could follow)?

danieldotnl commented:
> @danieldotnl, I addressed comments and I'm waiting for your guidance on some of them. Please review.
> 
> Also, I would love to add tests for these cases, so as they do not get broken over time. I tried adding a test and did not succeed after half an hour. I saw you doing stuff in that space recently. Do you plan to add tests (examples I could follow)?

Yes, I would love to add tests to multiscrape as well, but time is really limited for me at the moment. So let's not wait for that before merging :)

jeremicmilan commented on README.md:
```diff
@@ -88,7 +88,7 @@ Based on latest (pre) release.
 | authentication    | Configure HTTP authentication. `basic` or `digest`. Use this with username and password fields.                           | False    |         | string          |
```
I have some startup errors I need to investigate before completing this PR.

jeremicmilan replied:
@danieldotnl, the errors are due to configuration being incorrect, so please update that. I fixed it locally for me.

jeremicmilan commented:
@danieldotnl, I did final checks, and I'm fine for this to be merged (and of course, I'll add tests once you add some initial tests). So, please review, I'm happy addressing more issues (if any). :)
When merging, I suggest doing a squash merge (not sure how to specify that on GitHub, as this is my first PR here), as I have a bunch of small commits which would only pollute your main branch,

jeremicmilan commented:
@danieldotnl, can we finish the PR this week? From next week on, I'm again very tight on time (and it might be weeks before I'm able to address comments).

danieldotnl commented:
Jeremic, I was on vacation but good news! I tested the PR again and I'm happy with the changes! So I merged it and create a pre-release. Big thanks to you!!
Please install it in your production system and test it again.

There is one warning we still need to fix about a parser which is not specified. I think it's not provided to the form scraper as a parameter.

jeremicmilan commented:
@danieldotnl, thanks for merging and thanks for support on the PR!

I tested on my production server, and everything works great!

The only minor issue I have is that I need to scrape two different URLs that require the same form submit. As both logins go at the same time one of them is rejected by the server. The next retry starts working correctly. But... This is something I can live with. :)

What is the warning you are referring to? I'm not seeing it...

danieldotnl commented:
@jeremicmilan I'm planning to release this in a non-prerelease soon. Is it still working fine for you?
By the way, I fixed debugging and added the first automated tests!!

PS: You can find me also on discord if you want to chat. (instead of via this PR)

jeremicmilan commented:
Yep, works great. Feel free to publish.

I added a note to myself regarding adding tests for the changes I made. But I'm unsure when I'll have time to do this. 

Sure, I'll ping you once I start writing tests.
